README.md
:arrow_heading_up:
Go back to the
[Reference System Software repository](https://github.com/COPRS/reference-system-software)
:arrow_heading_up:

---

# Processing Sentinel 2

<!-- TOC -->
* [Processing Sentinel 2](#processing-sentinel-2)
  * [Overview](#overview)
  * [Available RS Addons](#available-rs-addons)
  * [Installation](#installation)
    * [Prerequisites](#prerequisites)
    * [Build](#build)
    * [Repository content](#repository-content)
    * [Using Ansible](#using-ansible)
    * [Manual Install](#manual-install)
    * [Uninstall](#uninstall)
* [Copyright and license](#copyright-and-license)
<!-- TOC -->

## Overview

This repository contains components for the Sentinel 2 processing chain.  
They are grouped by processing level, and deployed using a separate rs-addon for each level.

Each rs-addon includes : preparation worker, execution worker(s) and internal interface management
with rs-core components and Object Storage.

Basic workflow :  
![](inputs/media/s2_basic_workflow.png)

- The S2_L0U addon consumes Session data to produce basic datastrips/granules, SAD, and HKTM files.
- The S2_L0C addon uses the output of S2_L0U to produce L0 products.

For more information: [Sentinel 2 website](https://sentinels.copernicus.eu/web/sentinel/missions/sentinel-2).

## Available RS Addons

- [S2_L0U](rs-addons/S2_L0U)
- [S2_L0C](rs-addons/S2_L0C)

## Installation

Each RS-Addon will provide its own specific installation instructions, which may be found in their respective directory (see [above](#available-rs-addons)).

### Prerequisites

- Infrastructure : all the required tools (such as Kafka and MongoDB) are included in the RS infrastructure installation.  
  See  [Reference System Software Infrastructure](https://github.com/COPRS/infrastructure) for details.
- RS-Core : all the required components can be found in the following repositories :
    - [Production-Common](https://github.com/COPRS/production-common) (Ingestion, Catalog)
    - [Monitoring](https://github.com/COPRS/monitoring) (Processing monitoring & reporting)

### Build

In order to build the project from source, first clone the GitHub repository :

```shellsession
git clone https://github.com/COPRS/processing-sentinel-2.git
```

Then build the docker images:

```shellsession
./gradlew clean build bootBuildImage docker
```

And finally build the zip files:

```shellsession
./rs-addons/build_addons.sh
```

The zip files will be found in the rs-addons folder.

### Repository content

For each RS-Addon, the artifactory repository should contain:

- Docker images for the custom components of the addon in:  
  https://artifactory.coprs.esa-copernicus.eu/ui/repos/tree/General/rs-docker-private
- A zip file (its name includes the version number) for the addon in:  
  https://artifactory.coprs.esa-copernicus.eu/ui/repos/tree/General/rs-zip-private

### Using Ansible

Run the `deploy-rs-addon.yaml` playbook with the following variables:

- **stream_name**: name given to the stream in *Spring Cloud Dataflow*
- **rs_addon_location**: direct download url of the zip file or zip location on the bastion

Example:

```shellsession
ansible-playbook deploy-rs-addon.yaml \
    -i inventory/mycluster/hosts.yaml \
    -e rs_addon_location=https://artifactory.coprs.esa-copernicus.eu/artifactory/demo-zip/demo-rs-addon.zip \
    -e stream_name=example-stream-name
```

### Manual Install

Download and extract the zip file for the RS-Addon to install.  
If necessary, edit the parameters as required (See the specific addon release note for parameters description).

- Create all objects defined by files in _Executables/additional_resources_
- Using the SCDF GUI:
    - Register the applications using the content of the _stream-application-list.properties_ file
    - Create the streams using the content fo the _stream-definition.properties_ file
    - Deploy the stream using the properties defined in the _stream-parameters.properties_ file (removing comments)

### Uninstall

Using the SCDF GUI, undeploy then destroy all the streams relative to the RS-Addon.

# Copyright and license

The Reference System Software as a whole is distributed under the Apache License, version 2.0. A copy of this license is available in the [LICENSE](LICENSE) file. Reference System Software depends on third-party components and code snippets released under their own license (obviously, all compatible with the one of the Reference System Software). These dependencies are listed in the [NOTICE](NOTICE.md) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
> Content of release :
> - **Added** for new features.
> - **Changed** for changes in existing functionality.
> - **Deprecated** for soon-to-be removed features.
> - **Removed** for now removed features.
> - **Fixed** for any bug fixes.
> - **Security** in case of vulnerabilities.

## [1.8.0-rc1] - 2023-07-17

### Added
- Added limits and request values for filters and routers

### Changed

- Updated memory and cpu values
- Homogenization of the titles of the documentation

### Fixed

- Corrected Processing Station default value for L2 orchestrator
- [#999 - Added custom field to PW Tracelogs](https://github.com/COPRS/rs-issues/issues/999)

## [1.7.0-rc1] - 2023-05-17

### Added

-Updated IPF version for L1 processing to 6.1.4

### Changed

- Updated Spring Boot version

### Fixed

- [#935 - Update default size for s2-l1-shared volume](https://github.com/COPRS/rs-issues/issues/935)
- [#924 - Update default memory limits](https://github.com/COPRS/rs-issues/issues/924)
- [#928 - Use proper CatalogEvent fields to link GR/TL to their Datastrip](https://github.com/COPRS/rs-issues/issues/928)
- [#923 - Add timeout for obs transfers](https://github.com/COPRS/rs-issues/issues/923)
- [#901 - Remove RESTART allowed Action for L2PW](https://github.com/COPRS/rs-issues/issues/901)
- [#945 - Fixed Upload trigger when no file is produced](https://github.com/COPRS/rs-issues/issues/945)
- [#917 - Increase kafka timeout from 1 hour to 2 hours for L0C](https://github.com/COPRS/rs-issues/issues/917)
- [#941 - Updated MissingOutput behaviour](https://github.com/COPRS/rs-issues/issues/941)
- [#970 - Added Aux folder cleanup before download](https://github.com/COPRS/rs-issues/issues/970)

## [1.6.0-rc1] - 2023-03-20

### Added

- Rework docker build and add Security scan with Trivy
- Added upload of True Color Image files for L1C and L2A

### Changed

- Added copyrights to Notice and harmonized Readme titles

### Fixed

- [#807 - Added logic for L0u Processing in case no Datastrip is produced](https://github.com/COPRS/rs-issues/issues/807)
- [#856 - Increase direct memory size](https://github.com/COPRS/rs-issues/issues/856)
- [#863 - Rework and set proper allowed actions for ProcessingMessages](https://github.com/COPRS/rs-issues/issues/863)
- [#876 - Overwrite existing files during EW download](https://github.com/COPRS/rs-issues/issues/876)
- Set proper input for some traces

## [1.5.0-rc1] - 2023-01-26

### Added

- [#594 - Sentinel-2 Level-0 RS add-on to produce all the metadata files](https://github.com/COPRS/rs-issues/issues/594)
- [#683 - Update l0u and l0c docker image](https://github.com/COPRS/rs-issues/issues/683)

### Changed

- Remove unnecessary properties and set proper default for SAD Data

## [1.4.0-rc1] - 2023-01-04

### Added

- [#492 - Sentinel-2 Level-2 RS add-on](https://github.com/COPRS/rs-issues/issues/492)
- [#514 - Expose S2 pending processing as gauge metric](https://github.com/COPRS/rs-issues/issues/514)
- [#537 - Add rsChainVersion field in Processing Message](https://github.com/COPRS/rs-issues/issues/537)
- [#736 - Add ObsRead and ObsWrite traces](https://github.com/COPRS/rs-issues/issues/736)
- Misc documentation update

### Changed

- Update jdk version used by wrappers
- Update libraries versions

### Fixed

- [#734 - Set proper product family for SADATA](https://github.com/COPRS/rs-issues/issues/734)

## [1.3.0-rc1] - 2022-11-23

### Added

- [#573 - Sentinel-2 Level 1 RS addon Execution Worker Shared Context](https://github.com/COPRS/rs-issues/issues/573)
- [#574 - Sentinel-2 Level 1 RS addon Execution Worker L1AB](https://github.com/COPRS/rs-issues/issues/574)
- [#576 - Sentinel-2 Level 1 RS addon Preparation Worker L1C](https://github.com/COPRS/rs-issues/issues/576)
- [#575 - Sentinel-2 Level 1 RS addon Execution Worker L1C](https://github.com/COPRS/rs-issues/issues/575)
- Documentation for Level 1 RS addon

### Changed

- Rework some rs-addon configuration parameters

### Fixed

- [#654 - S2a L0c Execution fails](https://github.com/COPRS/rs-issues/issues/654)
- [#664 - L0u DS are missing in the END processing trace](https://github.com/COPRS/rs-issues/issues/664)

## [1.2.0-rc1] - 2022-10-25

### Added

- [#596 - Add missing namespace in the KafkaTopic additional resource](https://github.com/COPRS/rs-issues/issues/596)
- [#490 - Sentinel-2 Level-1 RS add-on Skeleton](https://github.com/COPRS/rs-issues/issues/490)
- [#546 - Set timeliness field for L0 RS add-on](https://github.com/COPRS/rs-issues/issues/546)
- [#572 - Sentinel-2 Level-1 RS add-on Preparation Worker](https://github.com/COPRS/rs-issues/issues/572)

## [1.0.0] - 2022-09-29

## [1.1.0-rc1] - 2022-09-27

### Added

- [#413 - Provide missing output on RS add-on S2 traces](https://github.com/COPRS/rs-issues/issues/413)

## [1.0.0-rc3] - 2022-09-22

### Added

- [#562 - Add checksum file next to S2 product for S3 uploads](https://github.com/COPRS/rs-issues/issues/562)
- Documentation and Notice.md

## [1.0.0-rc2] - 2022-09-02

### Fixed

- [#541 - Secrets are not encoded](https://github.com/COPRS/rs-issues/issues/541)
- [#542 - Double quotes in stream parameters](https://github.com/COPRS/rs-issues/issues/542)
- [#544 - Secrets use incorrect inventory variables](https://github.com/COPRS/rs-issues/issues/544)
- [#545 - Stream definition using labels raise warning message](https://github.com/COPRS/rs-issues/issues/545)
- fix traces format

## [1.0.0-rc1] - 2022-08-31

### Added

- Code Quality improvements
- Docs skeleton

## [0.11.0-rc3] - 2022-08-19

### Added

- [#446 - End User documentation](https://github.com/COPRS/rs-issues/issues/446)

## [0.11.0-rc2] - 2022-08-16

### Added

- [#450 - Trace ICD Compliance](https://github.com/COPRS/rs-issues/issues/450)

payloads/mdc2.sh
#safescale  gw-ivv-cluster  ~  curl -X GET http://10.98.60.209:9200/s2_aux/_search?pretty -H 'Content-type:application/json' -d
# '{"query":{"match":{"productType":"AUX_UT1UTC"}}}'

{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 4.6249723,
    "hits" : [
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2__OPER_AUX_UT1UTC_PDMC_20220407T000000_V20220408T000000_20230407T000000.TGZ",
        "_score" : 4.6249723,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2023-04-07T00:00:00.000000Z",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2022-04-07T00:00:00.000000Z",
          "insertionTime" : "2022-05-13T09:37:43.326886Z",
          "satelliteId" : "_",
          "validityStartTime" : "2022-04-08T00:00:00.000000Z",
          "productName" : "S2__OPER_AUX_UT1UTC_PDMC_20220407T000000_V20220408T000000_20230407T000000.TGZ",
          "productType" : "AUX_UT1UTC"
        }
      }
    ]
  }
}

# safescale  gw-ivv-cluster  ~  curl -X GET http://10.98.60.209:9200/s2_aux/_search?pretty

{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 152,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2DENT_MPC__20170206T103039_V20170101T000000_21000101T000000_B09.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "09",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:39.000000Z",
          "insertionTime" : "2022-05-13T09:37:38.713642Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_R2DENT_MPC__20170206T103039_V20170101T000000_21000101T000000_B09.TGZ",
          "productType" : "GIP_R2DENT"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2WAFI_MPC__20170206T103047_V20170101T000000_21000101T000000_B09.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "09",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:47.000000Z",
          "insertionTime" : "2022-05-13T09:37:40.972313Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_R2WAFI_MPC__20170206T103047_V20170101T000000_21000101T000000_B09.TGZ",
          "productType" : "GIP_R2WAFI"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2MACO_MPC__20170206T103040_V20170101T000000_21000101T000000_B09.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "09",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:40.000000Z",
          "insertionTime" : "2022-05-13T09:37:40.998560Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_R2MACO_MPC__20170206T103040_V20170101T000000_21000101T000000_B09.TGZ",
          "productType" : "GIP_R2MACO"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2DENT_MPC__20170206T103039_V20170101T000000_21000101T000000_B08.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "08",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:39.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.032179Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
            "productName" : "S2B_OPER_GIP_R2DENT_MPC__20170206T103039_V20170101T000000_21000101T000000_B08.TGZ",
          "productType" : "GIP_R2DENT"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_VIEDIR_MPC__20170512T114736_V20170322T000000_21000101T000000_B03.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "03",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-05-12T11:47:36.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.063324Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-03-22T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_VIEDIR_MPC__20170512T114736_V20170322T000000_21000101T000000_B03.TGZ",
          "productType" : "GIP_VIEDIR"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2L2NC_MPC__20170206T103040_V20170101T000000_21000101T000000_B04.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "04",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:40.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.107169Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_R2L2NC_MPC__20170206T103040_V20170101T000000_21000101T000000_B04.TGZ",
          "productType" : "GIP_R2L2NC"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2L2NC_MPC__20170206T103040_V20170101T000000_21000101T000000_B12.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "12",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:40.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.137357Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_R2L2NC_MPC__20170206T103040_V20170101T000000_21000101T000000_B12.TGZ",
          "productType" : "GIP_R2L2NC"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_R2DENT_MPC__20170206T103039_V20170101T000000_21000101T000000_B01.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "01",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:39.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.221261Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_R2DENT_MPC__20170206T103039_V20170101T000000_21000101T000000_B01.TGZ",
          "productType" : "GIP_R2DENT"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_VIEDIR_MPC__20170512T114736_V20170322T000000_21000101T000000_B02.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "02",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-05-12T11:47:36.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.249269Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-03-22T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_VIEDIR_MPC__20170512T114736_V20170322T000000_21000101T000000_B02.TGZ",
          "productType" : "GIP_VIEDIR"
        }
      },
      {
        "_index" : "s2_aux",
        "_type" : "_doc",
        "_id" : "S2B_OPER_GIP_TILPAR_MPC__20170206T103032_V20170101T000000_21000101T000000_B00.TGZ",
        "_score" : 1.0,
        "_source" : {
          "productFamily" : "S2_AUX",
          "validityStopTime" : "2100-01-01T00:00:00.000000Z",
          "bandIndexId" : "00",
          "missionId" : "S2",
          "productClass" : "OPER",
          "creationTime" : "2017-02-06T10:30:32.000000Z",
          "insertionTime" : "2022-05-13T09:37:41.274952Z",
          "satelliteId" : "B",
          "validityStartTime" : "2017-01-01T00:00:00.000000Z",
          "productName" : "S2B_OPER_GIP_TILPAR_MPC__20170206T103032_V20170101T000000_21000101T000000_B00.TGZ",
          "productType" : "GIP_TILPAR"
        }
      }
    ]
  }
}